File: create_bst.py
import random

class Node(object):
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None

	def __repr__(self):
		return "({})".format(self.data)

def create_bst(arr):
	l = len(arr)
	if l == 0:
		return None
	m = int(l / 2)
	n = Node(arr[m])
	n.left = create_bst(arr[:m])
	n.right = create_bst(arr[m + 1:])
	return n

def create_bst_optimized(arr, start = None, end = None):
	if start == None:
		start = 0
		end = len(arr) - 1
	if start > end:
		return None
	m = int((start + end) / 2)
	n = Node(arr[m])
	n.left = create_bst_optimized(arr, start, m - 1)
	n.right = create_bst_optimized(arr, m + 1, end)
	return n

def preorder_traverse(n):
	if not n:
		return
	preorder_traverse(n.left)
	print(n, end=", ")
	preorder_traverse(n.right)

def main():
	arr = [random.randint(100, 10000) for i in range(100)]
	arr = sorted(arr)
	print(arr)
	n = create_bst_optimized(arr)
	preorder_traverse(n)

if __name__ == "__main__":
	main()

File: create_bst.cpp
#include <iostream>
#include <cstdlib>

using namespace std;
#define NULL_VALUE -1

template <class T>
class Node{
public:
	Node();
	~Node();
	T data;
	Node<T>* left;
	Node<T>* right; 
};

template <class T>
Node<T>::Node(){
	left = NULL;
	right = NULL;
}

template <class T>
Node<T>::~Node(){
	delete data;
}

Node<int>* createBstOptimized(int* arr, int start = NULL_VALUE, int end = NULL_VALUE){
	// if(start == NULL_VALUE){
	// 	start = 0;
	// 	end = (sizeof(arr) / sizeof(int*)) -1;
	// 	cout << "end" << endl;
	// 	cout << sizeof(arr) << endl;
	// }
	if(start > end){
		return NULL;
	}
	int m = (start + end)/2;
	Node<int>* n = new Node<int>();
	n->data = arr[m];
	n->left = createBstOptimized(arr, start, m-1);
	n->right = createBstOptimized(arr, m+1, end);
	return n;
}

int preOrder(Node<int>* n){
	if(!n){
		return NULL_VALUE;
	}
	cout << n->data << endl;
	preOrder(n->left);
	preOrder(n->right);
}

int main(){

	int arr[10];
	for(int i=0; i < 10; i++){
		int random_integer = rand()%1000;
		arr[i] = random_integer;
		cout << arr[i] << endl;
	}

	int temp;
	for(int i=0;i< 10;i++){
		for(int j=0;j< 10-i;j++){
			if(arr[j]>arr[j+1]){
				temp=arr[j];				
				arr[j]=arr[j+1];
				arr[j+1]=temp;
			}
		}
	}
	
	cout << "sorted" << endl;
	for(int i=0; i < 10; i++){
		cout << arr[i] << endl;
	}

	Node<int>* n;
	n = createBstOptimized(arr, 0, 10);
	cout << "preOrder" << endl;
	preOrder(n);

	return 0;
}
